### PR TITLE
blob: release the adopted WTF string ref in string-body consumers

### PR DESCRIPTION
**Pull request by alii**
Consuming a string body via `Response`/`Request` `.json()` / `.arrayBuffer()` / `.bytes()` permanently leaked the body string. The Zig `defer str.deref()` that releases the AnyBlob's adopted +1 WTF::StringImpl ref was dropped from the three `Any` string-body arms in the Rust port (`bun_core::String` is `Copy` with no `Drop`, so an adopted ref returned without an explicit deref leaks unconditionally).

Wrap the adopted value in the existing `OwnedString` RAII helper at all three arms (`to_json`, `to_string`, `to_array_buffer_view`) — the Rust spelling of Zig's `defer str.deref()`, covering every return path. This is the same pattern already used elsewhere in `Blob.rs`.

Verified: the finding's `new Response(JSON.stringify(...)).json()` repro went from unbounded (~400 MB, linear, never plateaus) to bounded (plateaus, matching released Bun); `.arrayBuffer()`/`.bytes()` likewise. The regression test asserts boundedness via a two-block RSS growth check (absolute RSS deliberately not asserted).

Note: `.text()` is **not** claimed fixed here and is excluded from the test. While verifying I found a *separate, unrelated* regression — discarded large JS strings are not GC-reclaimed in this build independently of bodies (`Buffer.alloc(2e6).toString()` discarded in a loop leaks identically with no Blob/Response, while released Bun reclaims it). `.text()` returns a large string so its measurement is dominated by that separate bug; conflating them would test the wrong thing. Filing that separately.

### CODE REVIEW

**Comment by robobun** [comment]
🔧 Addressing review feedback — applying stderr capture, `describe.concurrent`, and verifying the `Buffer.alloc` suggestion does not reintroduce the separate large-string-reclaim issue noted in the test comment. Building...

File: test/js/web/fetch/body.test.ts [hunk 1]
```diff
@@ -668,3 +669,61 @@ function arrayBuffer(buffer: BufferSource) {
...
+  const cases: Array<[string, "Response" | "Request", string]> = [
+    ["Response.json", "Response", "json"],
+    ["Response.arrayBuffer", "Response", "arrayBuffer"],
+    ["Response.bytes", "Response", "bytes"],
+    ["Request.json", "Request", "json"],
+  ];
+
+  for (const [name, ctor, method] of cases) {
+    test(name, async () => {
+      const makeBody = method === "json" ? `() => JSON.stringify({ k: "z".repeat(SZ) })` : `() => "z".repeat(SZ)`;
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/web/fetch/body.test.ts]: Declining this one. The test's own comment (a few lines up) calls out `Buffer.alloc(2e6).toString()` as the trigger for a *separate* large-string-reclaim issue that is independent of the body-string deref being fixed here — switching the body generator to `Buffer.alloc(..).toString()` would conflate the two measurements. On top of that, `String#repeat` is measurably faster than `Buffer.alloc(..).toString()` on this path in the debug build (the `.toString()` decode dominates), so the perf premise doesn't hold here either. Noted inline in 102df1e9f1.

File: test/js/web/fetch/body.test.ts [hunk 2]
```diff
@@ -668,3 +669,61 @@ function arrayBuffer(buffer: BufferSource) {
...
+      });
+      const [stdout, exitCode] = await Promise.all([proc.stdout.text(), proc.exited]);
+      const m = stdout.match(/BLOCK1:(-?\d+) BLOCK2:(-?\d+)/);
+      expect(m).not.toBeNull();
+      const block2 = Number(m![2]);
+      // Bounded: BLOCK2 ≈ 0 (plateaued). Leaking: BLOCK2 ≈ 40 × ~3 MB ≈ 100+ MB
+      // and keeps growing every block. 50 MB cleanly separates the two.
+      expect(block2).toBeLessThan(50);
+      expect(exitCode).toBe(0);
+    });
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/web/fetch/body.test.ts]: Done in 102df1e9f1 — `stderr` is captured and asserted empty before the regex match.

File: test/js/web/fetch/body.test.ts [hunk 3]
```diff
@@ -668,3 +669,61 @@ function arrayBuffer(buffer: BufferSource) {
...
+        await run(BLOCK); Bun.gc(true); const c = rss();
+        console.log("BLOCK1:" + (b - a) + " BLOCK2:" + (c - b));
+      `;
+      await using proc = Bun.spawn({
+        cmd: [bunExe(), "-e", src],
+        env: bunEnv,
+        stdout: "pipe",
+        stderr: "pipe",
+      });
+      const [stdout, exitCode] = await Promise.all([proc.stdout.text(), proc.exited]);
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/web/fetch/body.test.ts]: Done in 102df1e9f1 — `stderr` is read and `expect(stderr).toBe("")` runs before the regex match so subprocess diagnostics surface first.

File: test/js/web/fetch/body.test.ts [hunk 4]
```diff
@@ -668,3 +669,61 @@ function arrayBuffer(buffer: BufferSource) {
...
 }
+
+// Consuming a string body via .text()/.json()/.arrayBuffer()/.bytes() used to
+// permanently leak the whole body string (the adopted WTF::StringImpl +1 ref
+// was never released), so RSS grew *linearly and never plateaued* across
+// round-trips. The fix mirrors the Zig `defer str.deref()`. This asserts the
+// leak is bounded: after a warmup, a second equal block of round-trips must
+// not keep growing RSS (a linear leak adds ~block-size every block; a bounded
+// impl plateaus). Absolute RSS is intentionally not asserted — only growth.
+describe("string body consumption does not leak", () => {
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/web/fetch/body.test.ts]: Done in 102df1e9f1 — switched to `describe.concurrent`. Also swapped the JSON arm's body to whitespace-padded JSON so the body string stays ~2 MB (what the leaked ref holds) but the parse is near-free; without that the JSON cases overran the default timeout under debug+ASAN once four subprocesses shared the CPU.

File: src/runtime/webcore/Blob.rs [hunk 5]
```diff
@@ -6438,8 +6442,12 @@ impl Any {
...
             }
             Any::WTFStringImpl(impl_) => {
-                let str =
-                    BunString::adopt_wtf_impl(core::mem::replace(impl_, core::ptr::null_mut()));
+                // Adopts a +1 WTF ref; `OwnedString` releases it on scope exit
+                // (Zig: `defer str.deref()`).
+                let str = OwnedString::new(BunString::adopt_wtf_impl(core::mem::replace(
+                    impl_,
+                    core::ptr::null_mut(),
+                )));
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/webcore/Blob.rs:6453]: Good catch — folded in as 6b738b2b43 via `self.detach()`.

One nuance on the repro: the *direct* `Response`/`Request` `.blob()` path goes through `Value::use_()` (Body.rs:1273-1292), whose `WTFStringImpl` arm already derefs before returning the new `Blob`, so `for(;;) await new Response(str).blob()` is actually bounded today (verified: BLOCK2 ≈ 2 MB on a build without this PR). `Any::to_blob` is reached from the stream-buffering action paths (`ByteStream`/`ByteBlobLoader` → `to_promise` → `to_action_value` with `BufferActionTag::Blob`), where the leak would surface. Fixing it here keeps all four `Any::WTFStringImpl` consumer arms consistent.
